### PR TITLE
[4.0]Fix component dispatcher loadLanguage() method

### PR DESCRIPTION
**Pull request by joomdonation**
Pull Request for Issue # .

### Summary of Changes
In case frontend code need to call a backend controller code (For example, com_config frontend call backend module controller save method to save module parameters), we will need some code like this (assume com_modules is already namespaced)

```php
JLoader::register('ModulesDispatcher', JPATH_ADMINISTRATOR . '/components/com_modules/dispatcher.php');

$app             = \Joomla\CMS\Application\CmsApplication::getInstance('administrator');
$dispatcher      = new ModulesDispatcher('Joomla\\Component\\Modules', $app, $input);
$controllerClass = $dispatcher->getController('Module');

// Execute backend controller
$return = $controllerClass->save();
```
When the dispatcher object is created, it call loadLanguage() https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/CMS/Dispatcher/Dispatcher.php#L93 method. The problem is that **$this->app->getLanguage()** return null and it causes fatal error ($language is only available in application object after calling **doExecute()** method). This PR proposes a fix for that issue

### Testing Instructions
This is code review only for Joomla 4 team

### CODE REVIEW

**Comment by wilsonge** [comment]
Is this not better?

```
JLoader::register('ModulesDispatcher', JPATH_ADMINISTRATOR . '/components/com_modules/dispatcher.php');

$app             = \Joomla\CMS\Application\CmsApplication::getInstance('administrator');
$app->loadLanguage($this->app->getLanguage());
$dispatcher      = new ModulesDispatcher('Joomla\\Component\\Modules', $app, $input);
$controllerClass = $dispatcher->getController('Module');

// Execute backend controller
$return = $controllerClass->save();
```

**Comment by joomdonation** [comment]
The problem is that when we create dispatcher object;
```php
$dispatcher      = new ModulesDispatcher('Joomla\\Component\\Modules', $app, $input);
```
The loadLanguage() method is called immediately https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/CMS/Dispatcher/Dispatcher.php#L93

The $app property of dispatcher object now is Administrator application object, it doesn't have $language property available yet, so $this->app->getLanguage() return null, thus it causes fatal error as I described

**Comment by joomdonation** [comment]
Ah, sorry. Maybe your code works  But in this case, language will be loaded twice I think. Will need to try it

File: libraries/src/CMS/Dispatcher/Dispatcher.php [hunk 1]
```diff
@@ -93,8 +93,10 @@ public function __construct($namespace, CMSApplication $app, \JInput $input = nu
 	protected function loadLanguage()
 	{
 		// Load common and local language files.
-		$this->app->getLanguage()->load($this->app->scope, JPATH_BASE, null, false, true) ||
-		$this->app->getLanguage()->load($this->app->scope, JPATH_COMPONENT, null, false, true);
+		$language = $this->app->getLanguage() ?: \JFactory::getLanguage();
+
+		$language->load($this->app->scope, JPATH_BASE, null, false, true) ||
+		$language->load($this->app->scope, JPATH_COMPONENT, null, false, true);
```
Comment (yvesh) [inline libraries/src/CMS/Dispatcher/Dispatcher.php]: Can we get rid of the app->scope and replace it through the option? The option can be easier injected with the input imo.

Comment (joomdonation) [inline libraries/src/CMS/Dispatcher/Dispatcher.php]: Yes. We need to do that. But later, once the autoload PR merged (if accepted). At that time, we need to review what need to be passed to dispatcher contructor..

**Comment by joomdonation** [comment]
OK. @wilsonge code works (although not easy to figure out). Let me know if you want to have the code like that and I will close this PR.

**Comment by wilsonge** [comment]
@mbabker thoughts please

**Comment by wilsonge** [comment]
My gut is that setting the existing language into the application is superior. But willing to listen to other arguments

**Comment by joomdonation** [comment]
BTW, loadLanguage() method in Dispatcher is not implemented properly. We are hardcode the path to JPATH_BASE and JPATH_COMPONENT which is wrong in the given sample code (the path must he JPATH_ADMINISTRATOR and JPATH_ADMINISTRATOR . '/components/com_modules';)

**Comment by wilsonge** [comment]
Agreed. I think JPATH_BASE is correct. But JPATH_COMPONENT is wrong

**Comment by joomdonation** [comment]
JPATH_BASE is not correct, too. With that sample code, the active component is com_config in the frontend, so JPATH_BASE equals JPATH_ROOT. But when we call com_modules backend, we need to language file from JPATH_ADMINISTRATOR, so JPATH_BASE is wrong

Unless application class has a method such as getPath() to return the root path of the application, I think we will need to use if else command here

File: libraries/src/CMS/Dispatcher/Dispatcher.php [hunk 2]
```diff
@@ -93,8 +93,10 @@ public function __construct($namespace, CMSApplication $app, \JInput $input = nu
 	protected function loadLanguage()
 	{
 		// Load common and local language files.
-		$this->app->getLanguage()->load($this->app->scope, JPATH_BASE, null, false, true) ||
-		$this->app->getLanguage()->load($this->app->scope, JPATH_COMPONENT, null, false, true);
+		$language = $this->app->getLanguage() ?: \JFactory::getLanguage();
```
Comment (laoneo) [inline libraries/src/CMS/Dispatcher/Dispatcher.php]: Why is that needed, is it ever possible that getLanguage() return null?

**Comment by wilsonge** [comment]
Ugh
